client/daemon/peer/traffic_shaper.go

```diff
@@ -0,0 +1,264 @@
+	ts.usingBandWidth.Add(int64(n))
+}
+
+func (ts *plainTrafficShaper) GetBandwidth() int64 {
+	return ts.usedBandwidth.Load()
+}
+
+type taskEntry struct {
+	ptc           *peerTaskConductor
+	pieceSize     uint32
+	usedBandwidth *atomic.Int64
+	needBandwidth int64
+	needUpdate    bool
+}
+
+type samplingTrafficShaper struct {
+	*logger.SugaredLoggerOnWith
+	sync.Mutex
+	computePieceSize func(int64) uint32
+	totalRateLimit   rate.Limit
+	usedBandwidth    *atomic.Int64
+	usingBandWidth   *atomic.Int64
+	tasks            map[string]*taskEntry
+	stopCh           chan struct{}
+}
+
+func NewSamplingTrafficShaper(totalRateLimit rate.Limit, computePieceSize func(int64) uint32) TrafficShaper {
+	log := logger.With("component", "TrafficShaper")
+	return &samplingTrafficShaper{
+		SugaredLoggerOnWith: log,
+		computePieceSize:    computePieceSize,
+		totalRateLimit:      totalRateLimit,
+		usedBandwidth:       atomic.NewInt64(0),
+		usingBandWidth:      atomic.NewInt64(0),
+		tasks:               make(map[string]*taskEntry),
+		stopCh:              make(chan struct{}),
+	}
+}
+
+func (ts *samplingTrafficShaper) Start() {
+	go func() {
+		// update bandwidth of all running tasks every second
+		ticker := time.NewTicker(time.Second)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				ts.usedBandwidth.Store(ts.usingBandWidth.Load())
+				ts.usingBandWidth.Store(0)
+				ts.updateLimit()
+			case <-ts.stopCh:
+				return
+			}
+		}
+	}()
+}
+
+func (ts *samplingTrafficShaper) Stop() {
+	close(ts.stopCh)
+}
+
+// updateLimit updates every task's limit every second
+func (ts *samplingTrafficShaper) updateLimit() {
+	var totalNeedBandwidth int64
+	var totalLeastBandwidth int64
+	ts.Lock()
+	defer ts.Unlock()
+	// compute overall remaining length of all tasks
+	for _, te := range ts.tasks {
+		var needBandwidth int64
+		oldLimit := int64(te.ptc.limiter.Limit())
+		needBandwidth = te.usedBandwidth.Load()
+		if !te.needUpdate {
+			// if this task is added within 1 second, don't reduce its limit this time
+			te.needUpdate = true
+			if oldLimit > needBandwidth {
+				needBandwidth = oldLimit
+			}
+		}
+		if contentLength := te.ptc.contentLength.Load(); contentLength > 0 {
+			remainingLength := contentLength - te.ptc.completedLength.Load()
+			if remainingLength < needBandwidth {
+				needBandwidth = remainingLength
+			}
+		}
+		// delta bandwidth
+		needBandwidth -= int64(te.pieceSize)
+		if needBandwidth < 0 {
+			needBandwidth = 0
+		}
+		te.needBandwidth = needBandwidth
+		totalNeedBandwidth += needBandwidth
+		totalLeastBandwidth += int64(te.pieceSize)
+		te.usedBandwidth.Store(0)
+	}
+
+	// allocate delta bandwidth for tasks
+	for _, te := range ts.tasks {
+		var deltaLimit float64
+		deltaLimit = (float64(ts.totalRateLimit) - float64(totalLeastBandwidth)) * (float64(te.needBandwidth) / float64(totalNeedBandwidth))
+		// make sure new limit is not smaller than pieceSize
+		if deltaLimit < 0 {
+			deltaLimit = 0
```

Comment:
If deltaLimit == 0, can we did not call `te.ptc.limiter.SetLimit`?

Reply:
`deltaLimit` indicates the difference between the allocated bandwidth and pieceSize, so if `deltaLimit` == 0, we still have to update the limit. The variable name is confusing, I have changed it to `diffLimit`